### PR TITLE
Only complete PipeReader/PipeWriter after operations complete

### LINKED ISSUE #350: HeaderDelimitedMessageHandler.Write throws when in race with disposal
If the `PipeMessageHandler` is writing (or reading) while being disposed, it can create a memory corruption and/or exception situation when the PipeWriter or PipeReader or their buffers are accessed after being completed.

VS has recently shown a tendency to fail with an infobar with this error:

```System.InvalidOperationException: Writing is not allowed after writer was completed.
 at System.IO.Pipelines.Pipe.GetMemory(Int32 sizeHint)
 at System.IO.Pipelines.Pipe.DefaultPipeWriter.GetMemory(Int32 sizeHint)
 at StreamJsonRpc.HeaderDelimitedMessageHandler.Write(JsonRpcMessage content, CancellationToken cancellationToken)
 at StreamJsonRpc.PipeMessageHandler.WriteCoreAsync(JsonRpcMessage content, CancellationToken cancellationToken)
 at StreamJsonRpc.MessageHandlerBase.<WriteAsync>d__15.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at StreamJsonRpc.JsonRpc.<TransmitAsync>d__118.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Threading.Tasks.ValueTask.ThrowIfCompletedUnsuccessfully()
 at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__102.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__96`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.VisualStudio.CodeLens.Proxy.CodeLensHubClient.<>c__DisplayClass22_0`1.<<InvokeRpcAsync>b__0>d.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.<CallExtensionPointAsync>d__36`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

### PR DESCRIPTION
**Pull request by AArnott**
We must not allow a race to exist between PipeReader/PipeWriter use and calling Complete() on them. This not only produces exceptions like the one in #350 but it opens up the possibility of accessing recycled buffers.

Fixes #350

This is the similar to #352 combined with #359, but targets v2.2 instead of v2.3.

### CODE REVIEW

**Comment by AArnott** [comment]
This suffers from *never* closing the pipe, since on .NET Framework the `NamedPipeClientStream.ReadAsync` call ignores cancellation tokens, so we must close the stream. This is (the forgotten reason) why #359 was important.

**Comment by AArnott** [comment]
I've pushed a change that addresses it almost the same way that we fix it for v2.3. The only difference being we hide the added APIs and don't call DisposeAsync on the MessageHandlerBase.

File: src/StreamJsonRpc/MessageHandlerBase.cs [hunk 1]
```diff
@@ -35,6 +35,26 @@ public abstract class MessageHandlerBase : IJsonRpcMessageHandler, IDisposableOb
...
+        /// </summary>
+        private readonly AsyncManualResetEvent readingCompleted = new AsyncManualResetEvent();
+
+        /// <summary>
+        /// A signal that the last write operation has completed.
+        /// </summary>
+        private readonly AsyncManualResetEvent writingCompleted = new AsyncManualResetEvent();
+
+        /// <summary>
+        /// A value indicating whether the <see cref="ReadAsync(CancellationToken)"/> method is in progress.
```
Comment (ikeras) [inline src/StreamJsonRpc/MessageHandlerBase.cs]: The comment seems a little off, as this state indicates whether a ReadAsync and/or a WriteAsync method is in progress

Comment (AArnott) [reply src/StreamJsonRpc/MessageHandlerBase.cs]: Fixed

**Pull request merged**